File: frontend/lib/api/vehicles.ts
import { api } from "@/lib/api/client";
import type {
  VehicleCreate,
  VehicleDeleteResponse,
  VehicleResponse,
  VehicleStatus,
  VehicleUpdate,
} from "@/types/vehicles";

// `capacity` is a Decimal on the backend and serializes as a JSON string,
// not a number - normalize it here so arithmetic on it (LC gauges in the
// Loading Supervisor module) doesn't silently do string concatenation.
function normalize(vehicle: VehicleResponse): VehicleResponse {
  return { ...vehicle, capacity: Number(vehicle.capacity) };
}

export function listVehicles() {
  return api.get<VehicleResponse[]>("/vehicles").then((res) => res.data.map(normalize));
}

export function createVehicle(data: VehicleCreate) {
  return api.post<VehicleResponse>("/vehicles", data).then((res) => normalize(res.data));
}

export function updateVehicle(vehicleId: string, data: VehicleUpdate) {
  return api.patch<VehicleResponse>(`/vehicles/${vehicleId}`, data).then((res) => normalize(res.data));
}

export function assignVehicleDriver(vehicleId: string, driverId: string) {
  return api
    .patch<VehicleResponse>(`/vehicles/${vehicleId}/driver`, { driver_id: driverId })
    .then((res) => normalize(res.data));
}

export function setVehicleStatus(vehicleId: string, status: VehicleStatus) {
  return api
    .patch<VehicleResponse>(`/vehicles/${vehicleId}/status`, { status })
    .then((res) => normalize(res.data));
}

export function deleteVehicle(vehicleId: string) {
  return api.delete<VehicleDeleteResponse>(`/vehicles/${vehicleId}`).then((res) => res.data);
}

File: backend/app/api/vehicles.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.core.deps import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.vehicle import (
    VehicleCreate,
    VehicleUpdate,
    VehicleDriverUpdate,
    VehicleStatusUpdate,
    VehicleResponse,
    VehicleDeleteResponse,
)
from app.services import vehicle as vehicle_service
from app.services.vehicle import DuplicateVehicleError

router = APIRouter(prefix="/vehicles", tags=["vehicles"])


@router.post("", response_model=VehicleResponse, status_code=status.HTTP_201_CREATED)
def create_vehicle(
    data: VehicleCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        return vehicle_service.create_vehicle(db, data)
    except DuplicateVehicleError as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e))


@router.patch("/{vehicle_id}", response_model=VehicleResponse)
def update_vehicle(
    vehicle_id: uuid.UUID,
    data: VehicleUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        vehicle = vehicle_service.update_vehicle(db, vehicle_id, data)
    except DuplicateVehicleError as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e))

    if vehicle is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Vehicle not found")
    return vehicle


@router.patch("/{vehicle_id}/driver", response_model=VehicleResponse)
def assign_driver(
    vehicle_id: uuid.UUID,
    data: VehicleDriverUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    vehicle = vehicle_service.assign_driver(db, vehicle_id, data.driver_id)
    if vehicle is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Vehicle not found")
    return vehicle


@router.patch("/{vehicle_id}/status", response_model=VehicleResponse)
def update_vehicle_status(
    vehicle_id: uuid.UUID,
    data: VehicleStatusUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    vehicle = vehicle_service.set_vehicle_status(db, vehicle_id, data.status)
    if vehicle is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Vehicle not found")
    return vehicle


@router.delete("/{vehicle_id}", response_model=VehicleDeleteResponse)
def delete_vehicle(
    vehicle_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    vehicle = vehicle_service.soft_delete_vehicle(db, vehicle_id)
    if vehicle is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Vehicle not found")
    return vehicle
